internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -5,61 +5,84 @@
-	UploadFn   UploadFunc // Custom implementation on how to upload the contents under a file. Defaults to S3UploadFn.
-	Recursive  bool       // Whether to walk recursively.
+	// Upload is the function called when uploading a file.
+	Upload func(path string, contents io.Reader) (string, error)
+
+	// PathPrefix is the path to prefix any hashed files when uploading to the cache.
+	PathPrefix string
+
+	// AssetMappingPath is the path the upload the asset mapping file to.
+	AssetMappingPath string
 }
 
-// Upload uploads static assets to Cloud Storage and returns uploaded file URLs.
-func Upload(fs afero.Fs, source, destination string, opts *UploadOpts) ([]string, error) {
-	matcher := buildCompositeMatchers(buildReincludeMatchers(opts.Reincludes), buildExcludeMatchers(opts.Excludes))
-	var urls []string
-	info, err := fs.Stat(source)
-	if err != nil {
-		return nil, fmt.Errorf("get stat for file %q: %w", source, err)
-	}
-	paths := []string{source}
-	if info.IsDir() {
-		files, err := afero.ReadDir(fs, source)
-		if err != nil {
-			return nil, fmt.Errorf("read directory %q: %w", source, err)
-		}
-		paths = make([]string, len(files))
-		for i, f := range files {
-			paths[i] = filepath.Join(source, f.Name())
+type cached struct {
+	localPath string
+	content   io.Reader
+
+	Path            string `json:"path"`
+	DestinationPath string `json:"dest_path"`
+}
+
+// UploadFiles hashes each of the files specified in files and uploads
+// them to the path "{CachePathPrefix}/{hash}". After, it uploads a JSON file
+// to CacheMovePath that specifies the uploaded location of every file and it's
+// intended destination path.
+func (u *CacheUploader) UploadFiles(files []manifest.FileUpload) (string, error) {
+	var assets []cached
+	for _, f := range files {
+		matcher := buildCompositeMatchers(buildReincludeMatchers(f.Reinclude.ToStringSlice()), buildExcludeMatchers(f.Exclude.ToStringSlice()))
+		source := filepath.Join(f.Context, f.Source)
+
+		if err := afero.Walk(u.FS, source, u.walkFn(source, f.Destination, f.Recursive, matcher, &assets)); err != nil {
+			return "", fmt.Errorf("walk the file tree rooted at %q: %w", source, err)
 		}
-	} else if destination == "" { // only applies to files, not folders
-		destination = source
 	}
-	for _, path := range paths {
-		if err := afero.Walk(fs, path, walkFn(source, destination, opts.Recursive, fs, opts.UploadFn, &urls, matcher)); err != nil {
-			return nil, fmt.Errorf("walk the file tree rooted at %q: %w", source, err)
-		}
+
+	if err := u.uploadAssets(assets); err != nil {
+		return "", err
+	}
+
+	mappingFile := &bytes.Buffer{}
+	if err := json.NewEncoder(mappingFile).Encode(assets); err != nil {
+		return "", fmt.Errorf("unable to encode move json")
 	}
-	return urls, nil
+
+	// upload move file
+	url, err := u.Upload(u.AssetMappingPath, mappingFile)
+	if err != nil {
+		return "", fmt.Errorf("unable to upload cache move: %w", err)
+	}
+
+	return url, nil
```

Comment:
does the URL need to be returned or can we get it from either of these:
https://github.com/aws/copilot-cli/blob/d85cbbe5e36a7810bbe4342e2a121692719ac5bd/internal/pkg/aws/s3/s3.go#L158-L171